### DISCUSSION TITLE
Create a tag when merge happens to master branch

### DISCUSSION THREAD

**Discussion by NaveenKundeti-IBS** [community/community#54136] [Actions]
### Select Topic Area

Bug

### Body

**Working scenario:** When I merge the PR by clicking the "Merge pull request" button on PR the `push` event action runs my tag workflow and creates a tag from the master.
**Not working scenario:** When I merge PR by using "slash comments" ex: /Approved the PR is merging but the tag workflow is not triggering.

Please give me a solution. Why is it not triggering?
What was the difference by using the "Merge pull request" button and "slash comments" for merge?

**Comment by mohammad-zolghadr** [comment]
The reason why the tag workflow is not triggering when merging the PR using slash commands (/Approved) instead of the "Merge pull request" button is that the push event is not triggered when using the slash command to merge. Instead, the GitHub API is used to perform the merge operation directly on the server-side, which does not generate a push event that triggers the tag workflow.

To trigger the tag workflow when using slash commands to merge, you can modify your workflow to listen for the `pull_request` event instead of the `push` event. This event will be triggered whenever a pull request is opened, reopened, synchronized, or closed. You can then check the state of the pull request to determine if it has been merged, and if so, create a tag from the base branch.

The difference between using the "Merge pull request" button and slash commands to merge is that the former generates a push event that triggers the tag workflow, while the latter does not. Additionally, using slash commands can be more convenient for quick merges, but it may not provide the same level of control and visibility as merging through the GitHub UI.

**Reply by NaveenKundeti-IBS** [reply]
Thaks @mohammad-zolghadr. 
Yes, My workflow will listen to both events like below: 
`on:
  pull_request:
    types: [closed]
  push:
    branches:
      - master`
      
But the problem here is my other merge workflow has the event "issue_comment" action for merge PR using the `slash command`. After that The PR will be merged and closed also all the commits will be merged to the `master` branch as the same as "Merge pull request". But the difference only I found is the user by using slash comments "github-actions[bot]" will merge the PR is that the issue?
![image](https://user-images.githubusercontent.com/98738853/235427360-8e51da78-8843-4df1-a16b-c6a94a3333b1.png)

However, We are committing or pushing something to the master branch so the actions should run right?

**Reply by mohammad-zolghadr** [reply]
If you have configured your workflow to listen for both the `pull_request` and `push` events, and the workflow is still not triggering when you merge a pull request using slash commands, then there may be another issue with your workflow configuration.

One possible issue is that the `issue_comment` event may be triggering a separate workflow that does not include the steps to create a tag from the master branch. In this case, you may need to modify the `issue_comment` workflow to include the necessary steps to create the tag.

It's also possible that there may be a permissions issue preventing the `github-actions[bot]` user from triggering the workflow when merging with slash commands. You may want to double-check the permissions and settings for your repository to ensure that the bot user has the necessary permissions to trigger the workflow.

If you are committing or pushing something to the master branch, the actions should run, but it's possible that there may be a configuration issue preventing the workflow from running correctly. You may want to check the logs and status of the workflow to see if there are any error messages or warnings that can help identify the issue.

**Comment by cyrusjameskhan** [comment]
Difference between "Merge pull request" button and "slash comments":

1. "Merge pull request" button: This is the standard method provided by GitHub for merging a pull request. When you click this button, it will merge the changes, close the PR, and trigger any configured GitHub Actions, such as your tag workflow.
2. "slash comments": This method involves using specific commands in the PR comments section (e.g., /Approved) to perform actions. These commands are not native to GitHub, but rather, they are part of a third-party tool or bot that has been integrated with your repository. This integration may not trigger GitHub Actions by default when a PR is merged.

Possible solutions:

1. Ensure that the third-party tool or bot you're using to merge PRs with "slash comments" is configured to trigger GitHub Actions. You may need to consult the documentation of the tool or bot you're using, or contact its support team for assistance.

2. Consider using a different bot or integration that supports triggering GitHub Actions when merging PRs using "slash commands."

3. If you don't specifically require "slash comments" for merging PRs, you can stick with the standard "Merge pull request" button, which triggers your tag workflow as expected.

**Comment by hrishikesh-kocharekar-dev** [comment]
The reason why your tag workflow is not triggering when you merge a pull request using slash commands is likely because the slash command does not trigger the same events as the "Merge pull request" button.

When you click the "Merge pull request" button, it triggers a push event that includes the merged changes into the master branch, which then triggers your tag workflow. However, when you use a slash command to merge a pull request, it does not trigger a push event to the master branch, and therefore your tag workflow is not triggered.

One solution to this issue is to modify your tag workflow to trigger on the creation of a pull request instead of a push event to the master branch. This way, your workflow will be triggered regardless of how the pull request is merged.

To modify your tag workflow to trigger on the creation of a pull request, you can use the `pull_request` event in your workflow file. Here's an example:
on:
  pull_request:
    types: [closed]

jobs:
  create-tag:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v2
      - name: Create tag
        run: |
          git tag -a v1.0 -m "Version 1.0"
          git push --follow-tags

In this example, the `pull_request` event is used to trigger the workflow when the pull request is closed (which includes merging). The `create-tag` job then checks out the code, creates a tag, and pushes the tag to the repository.

By using this method, your tag workflow will be triggered whenever a pull request is merged, regardless of whether it was merged using the "Merge pull request" button or a slash command.

**Reply by NaveenKundeti-IBS** [reply]
I tried this `pull_request:
types: [closed]` but even it is not triggered after the pull request is closed. Actually, after I use `/Approved` it will merge my PR using PyGithub API `( pr.merge(merge_method = 'merge', commit_message ='Pull Request Approved and Merged!'))`. Should I change the merge method or is there anything causing this issue?
